lib/preview/previewUtils.js

```diff
@@ -0,0 +1,183 @@
+const { getAccountId, isTrackingAllowed, getAccountConfig } = require('../config');
+const { platform, release } = require('os');
+const { trackUsage } = require('../../api/fileMapper');
+const unAuth = require('../../api/localDevAuth/unauthenticated');
+
+const VALID_PROXY_DOMAIN_SUFFIXES = ['localhost', 'hslocal.net'];
+const HS_PREVIEW_GATE = "cms:localHublPreviews";
+
+const getPortalDomains = async (portalId) => {
+  try {
+    const result = await fetchDomains(portalId);
+    return result;
+  } catch (error) {
+    console.log("There was a problem fetching domains for your portal. You may be missing a scope necessary for this feature.")
+    return [];
+  }
+}
+
+const getPreviewUrl = (sessionInfo, queryParams) => {
+  const { portalId, env, hublet } = sessionInfo;
+
+  return `http://${portalId}.hubspotpreview${
+    env === 'qa' ? 'qa' : ''
+  }-${hublet}.com/_hcms/preview/template/multi?${stringifyQuery(queryParams)}`;
+}
+
+const stringifyQuery = (query) => {
+  return Object.keys(query)
+    .sort()
+    .map(key => `${key}=${query[key]}`)
+    .join('&');
+}
+
+const insertAtEndOfBody = (html, script) => {
+  const insertAt = (baseStr, index, insertStr) => {
+    return `${baseStr.slice(0, index)}${insertStr}${baseStr.slice(index)}`;
+  }
+  const endOfBodyIndex = html.lastIndexOf("</body>");
+  return insertAt(html, endOfBodyIndex, script);
+}
+
+const addRefreshScript = (html) => {
+  const refreshScript = `
+  <script>
+  (() => {
+    const NORMAL_WAIT_MS = 1000;
+    const BACKOFF_RATIO = 2;
+    let nextWait = NORMAL_WAIT_MS;
+    let timeWaited = 0;
+
+    setInterval(() => {
+      timeWaited += NORMAL_WAIT_MS;
+      if (timeWaited !== nextWait) return;
+      timeWaited = 0;
+      const res = fetch(window.location.origin + '/meta')
+        .then(async (res) => {
+          const hsServerState = await res.json();
+          if (hsServerState["REMOTE_FS_IS_DIRTY"]) {
+            location.reload();
+          }
+          nextWait = NORMAL_WAIT_MS;
+        })
+        .catch(err => {
+          nextWait *= BACKOFF_RATIO;
+          console.log('Disconnected from local server... (retrying in ' + nextWait / 1000 + 's)');
+        });
+    }, NORMAL_WAIT_MS);
+  })();
+  </script>
+  `;
+  return insertAtEndOfBody(html, refreshScript);
+}
+
+const getSubDomainFromValidLocalDomain = hostname => {
+  for (const validProxyDomainSuffix of VALID_PROXY_DOMAIN_SUFFIXES) {
+    if (hostname.endsWith(`.${validProxyDomainSuffix}`)) {
+      return hostname.slice(0, -1 * validProxyDomainSuffix.length - 1);
+    }
+  }
+  return undefined;
+};
+
+// From(ish) https://git.hubteam.com/HubSpot/cloudflare-workers/blob/master/worker-lib/src/Constants.ts
+const internalRoutes = {
+  HCMS: '/_hcms/',
+  HS_FS: '/hs-fs/',
+  HUB_FS: '/hubfs/'
+}
+
+const isInternalCMSRoute = (req) =>
+  Object.values(internalRoutes).some((route => req.path.startsWith(route)));
+
+const silenceConsoleWhile = async (act, ...args) => {
```

Comment:
I mentioned it elsewhere, but make sure to use `logger` everywhere. Seems reasonable for us to support some sort of `silenced` status on the logger. Might be as easy as setting `LOG_LEVEL` to `NONE`.

Reply:
A little wonky - the library I'm using here `mkcert-cli` works great for our use case [(more on other options here)](https://github.com/HubSpot/cli-lib/pull/32#discussion_r1374996546) of replicating the `vite-plugin-mkcert` behavior that the JSR  server has - but it uses console.log every function with no way to turn it off - so this is a util function to overwrite the console object temporarily while the cert generation runs synchronously and then put it back. 

It's only available as an ESM module so had to dynamically import it - once I move this to local-dev-lib I should be able to do the import normally at the top of the file